src/common/type/pack.c

```diff
@@ -0,0 +1,1513 @@
+
+OBJECT_DEFINE_FREE(PACK_WRITE);
+
+/**********************************************************************************************************************************/
+// Helper to create common data
+static PackRead *
+pckReadNewInternal(void)
+{
+    FUNCTION_TEST_VOID();
+
+    PackRead *this = NULL;
+
+    MEM_CONTEXT_NEW_BEGIN("PackRead")
+    {
+        this = memNew(sizeof(PackRead));
+
+        *this = (PackRead)
+        {
+            .memContext = MEM_CONTEXT_NEW(),
+            .tagStack = lstNewP(sizeof(PackTagStack)),
+        };
+
+        this->tagStackTop = lstAdd(this->tagStack, &(PackTagStack){.type = pckTypeObj});
+    }
+    MEM_CONTEXT_NEW_END();
+
+    FUNCTION_TEST_RETURN(this);
+}
+
+PackRead *
+pckReadNew(IoRead *read)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(IO_READ, read);
+    FUNCTION_TEST_END();
+
+    ASSERT(read != NULL);
+
+    PackRead *this = pckReadNewInternal();
+    this->read = read;
+    this->buffer = bufNew(ioBufferSize());
+    this->bufferPtr = bufPtr(this->buffer);
+
+    FUNCTION_TEST_RETURN(this);
+}
+
+PackRead *
+pckReadNewBuf(const Buffer *buffer)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(BUFFER, buffer);
+    FUNCTION_TEST_END();
+
+    ASSERT(buffer != NULL);
+
+    PackRead *this = pckReadNewInternal();
+    this->bufferPtr = bufPtrConst(buffer);
+    this->bufferMax = bufUsed(buffer);
+
+    FUNCTION_TEST_RETURN(this);
+}
+
+/***********************************************************************************************************************************
+Read bytes from the buffer
+
+IMPORTANT NOTE: To avoid having dyamically created return buffers the current buffer position (this->bufferPos) is stored in the
+object. Therefore this function should not be used as a parameter in other function calls since the value of this->bufferPos will
+change.
+***********************************************************************************************************************************/
+static size_t pckReadBuffer(PackRead *this, size_t size)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_READ, this);
+        FUNCTION_TEST_PARAM(SIZE, size);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+
+    size_t remaining = this->bufferMax - this->bufferPos;
+
+    if (remaining < size)
+    {
+        if (this->read != NULL)
+        {
+            // Nothing can be remaining since each read fetches exactly the number of bytes required
+            ASSERT(remaining == 0);
+            bufUsedZero(this->buffer);
+
+            // Limit the buffer for the next read so we don't read past the end of the pack
+            bufLimitSet(this->buffer, size < bufSizeAlloc(this->buffer) ? size : bufSizeAlloc(this->buffer));
+
+            // Read bytes
+            ioReadSmall(this->read, this->buffer);
+            this->bufferPos = 0;
+            this->bufferMax = bufUsed(this->buffer);
+            remaining = this->bufferMax - this->bufferPos;
```

Comment:
`remaining` is always going to be equal to bufferMax since bufferPos is always reset to 0. Why is that?

Reply:
Good point --  probably copied that from above. Fixed in 84128037a1a.